### ISSUE TITLE
Permissions - can I black list?

### ISSUE THREAD

**Issue by kyleboyle** [janino-compiler/janino#44]
I'm finding the documentation on permissions hard to follow. Is it true that if I don't call `setPermissions` then all permissions are enabled by default? If I call `setPermissions` with a permission list, only those permissions are enabled. Is it possible to approach from the other angle and  disallow certain permissions?

**Comment by kyleboyle** [comment]
Thanks for the response. Do you think there would be a reliable way to iterate through all permissions and exclude the permissions and exclude the ones to blacklist?

**Comment by aunkrig** [comment]
Puh... Maybe this helps?

https://docs.oracle.com/javase/tutorial/security/index.html

**Comment by aunkrig** [comment]
The question is: Why do you want to blacklist? What is your use case?

**Comment by aunkrig** [comment]
This seems to do what you want:

````java
import java.security.AllPermission;
import java.security.Permission;
import java.security.PermissionCollection;
import java.security.Permissions;
import java.util.Enumeration;
import java.util.PropertyPermission;

import org.codehaus.commons.compiler.Sandbox;

public class Foo {

    public static void main(String[] args) {
        
        Permission           p  = new PropertyPermission("B", "read");
        PermissionCollection pc = new Permissions();
        pc.add(p);
        
        pc = new InvertedPermissionCollection(pc);
        
        Sandbox.confine(Foo.class.getClassLoader(), pc);
        
        System.getProperty("A"); // Works OK.
        System.getProperty("B"); // java.security.AccessControlException: access denied ("java.util.PropertyPermission" "B" "read")
    }

    static final
    class InvertedPermissionCollection extends PermissionCollection {

        private static final long serialVersionUID = 1L;
        
        private final PermissionCollection delegate;

        public
        InvertedPermissionCollection(PermissionCollection delegate) { this.delegate = delegate; }

        @Override public boolean
        implies(Permission permission) { return !delegate.implies(permission); }

        @Override public void                    add(Permission permission) { throw new UnsupportedOperationException(); }
        @Override public Enumeration<Permission> elements()                 { throw new UnsupportedOperationException(); }
    }
}
``

**Comment by kyleboyle** [comment]
Arno, clever solution, thanks so much for looking into this!
FYI - my use case is allowing a user to create scripts to run in my system, but disallow certain potentially dangerous operations like starting threads or opening ports.
